File: ObjectOrientedPatterns/AbstractFactory.cs
namespace ObjectOrientedPatterns
{
    namespace AbstractFactory
    {
        public interface ICheapCar
        {
            public void Drive();
        }

        public interface IExpensiveCar
        {
            public void Drive();
            public void Entertainment();
        }

        public interface ICarFactory
        {
            public ICheapCar CreateCheapCar();
            public IExpensiveCar CreateExpensiveCar();
        }

        internal class CheapOpel : ICheapCar
        {
            public void Drive()
            {
                Console.WriteLine("driving a cheap Opel");
            }
        }

        internal class ExpensiveOpel : IExpensiveCar
        {
            public void Drive()
            {
                Console.WriteLine("driving an expensive Opel");
            }

            public void Entertainment()
            {
                Console.WriteLine("Opel entertainment");
            }
        }

        internal class CheapTesla : ICheapCar
        {
            public void Drive()
            {
                Console.WriteLine("driving a cheap Tesla (you're walking)");
            }
        }

        internal class ExpensiveTesla : IExpensiveCar
        {
            public void Drive()
            {
                Console.WriteLine("driving an expensive Tesla");
            }

            public void Entertainment()
            {
                Console.WriteLine("Tesla entertainment");
            }
        }

        public class OpelFactory : ICarFactory
        {
            public ICheapCar CreateCheapCar()
            {
                return new CheapOpel();
            }

            public IExpensiveCar CreateExpensiveCar()
            {
                return new ExpensiveOpel();
            }
        }

        public class TeslaFactory : ICarFactory
        {
            public ICheapCar CreateCheapCar()
            {
                return new CheapTesla();
            }

            public IExpensiveCar CreateExpensiveCar()
            {
                return new ExpensiveTesla();
            }
        }
    }
}

File: FunctionalPatterns/Creational/AbstractFactory.fs
namespace FunctionalPatterns
namespace FunctionalPatterns.Creational

module AbstractFactory =

    // Observation: Record types that just hold functions can be used like interfaces

    type CheapCar = { drive: unit -> unit }

    type ExpensiveCar =
        { drive: unit -> unit
          entertain: unit -> unit }

    type CarFactory =
        { createCheapCar: unit -> CheapCar
          createExpensiveCar: unit -> ExpensiveCar }

    let private newCeapOpel () =
        { CheapCar.drive = fun () -> printfn "driving a cheap Opel" }

    let private newExpensiveOpel () =
        { ExpensiveCar.drive = fun () -> printfn "driving an expensive Opel"
          entertain = fun () -> printfn "Opel entertainment" }

    let private newCheapTesla () =
        { CheapCar.drive = fun () -> printfn "driving a cheap Tesla (you're walking)" }

    let private newExpensiveTesla () =
        { ExpensiveCar.drive = fun () -> printfn "driving an expensive Tesla"
          entertain = fun () -> printfn "Tesla entertainment" }

    let newOpelFactory () =
        { createCheapCar = fun () -> newCeapOpel ()
          createExpensiveCar = newExpensiveOpel }

    let newTeslaFactory () =
        { createCheapCar = newCheapTesla
          createExpensiveCar = newExpensiveTesla }
